### ISSUE TITLE
Identify and document internal code used by providers

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFrameworkCore#11266]
Following up on the discussion from #11132, we have decided to take the following action for 2.1:
* Identify the internal types that are used by the "big five" providers (Postgres, MySQL, SQLite, SQL Server, SQL Compact) and add a note to each of these types to avoid breaking in the 2.1 timeframe
* File an issue for each class of internal code--e.g. one for resource strings, one for conventions, and so on. For the most part we will address these things post 2.1.
  * If there is already a way to do something without using internals, then file an issue on the relevant providers instead.

**Comment by ajcvickers** [comment]
@ErikEJ @roji @caleblloyd Would appreciate your thoughts on this.

**Comment by roji** [comment]
@ajcvickers would you like me to open issues for the points raised in #11132?

**Comment by ajcvickers** [comment]
@roji That would be great, especially if you could include the actual internal types that are being used so we know where to go make annotations in our code.

**Comment by smitpatel** [comment]
Pomelo.MySql

Annotations Framework
	- InternalEntityTypeBuilder
	- RelationalAnnotationsBuilder
	- ConfigurationSource
	- InternalIndexBuilder
	- InternalModelBuilder
	- InternalPropertyBuilder
	- InternalKeyBuilder
	- InternalRelationalBuilder
	- RelationalForeignKeyBuilderAnnotations
	
Conventions
	- DatabaseGeneratedAttributeConvention
	- IModelInitializedConvention
	- RelationalConventionSetbuilder
	- RelationalConventionSetBuilderDependencies
	- CoreConventionSetBuilder
	- CoreConventionSetBuilderDependencies
	- IConventionSetBuilder

RevEng
	- TableSelectionSet

Logging
	- RelationalStrings
	- CoreStrings

Utility
	- DisplayName()
	- ShortDisplayName()

Command
	- IndentedStringBuilder
	- RelationalParameterBuilder
	- RelationalCommand
	- Logger.CommandExecuting/Executed/Error log mssages
	- TransactionLogger.TransactionStarted/Used/Committed/Error/Rolledback

**Comment by smitpatel** [comment]
Npgsql

Annotations Framework
	- InternalEntityTypeBuilder
	- ConfigurationSource
	- InternalIndexBuilder
	- InternalModelBuilder
	- InternalPropertyBuilder
	- InternalKeyBuilder
	- InternalRelationalBuilder
	- RelationalForeignKeyBuilderAnnotations
	- RelationalIndexBuilderAnnotations
	- RelationalKeyBuilderAnnotations
	- RelationalEntityTypeBuilderAnnotations
	- RelationalAnnotationsBuilder
	- Property
	
Conventions
	- IModelInitializedConvention
	- RelationalConventionSetbuilder
	- RelationalConventionSetBuilderDependencies
	- RelationalMaxIdentifierLengthConvention
	- ICoventionSetBuilder

Query
	- FindProperty(this expression, type)
	- MemberAccessBindingExpressionVisitor.GetPropertyPath
	- QueryCompilationContextDependencies
	- ILinqOperatorProvider
	- LinqOperatorProvider
	- AsyncLinqOperatorProvider

Logging
	- RelationalStrings
	- CoreStrings

Utility
	- DisplayName()
	- ShortDisplayName()

**Comment by smitpatel** [comment]
SqlCe/SqlServer/SQLite is hard to figure out at present because they use the same namespace as EFCore hence no way to differentiate between EFCore type vs provider type.

Blocked on #11132

**Comment by roji** [comment]
Thanks for the legwork @smitpatel. Just to say that although this is important, it isn't incredibly urgent for me - in general I try to keep up with EF Core so Internal changes at your end are usually a minor annoyance more than anything else.

**Comment by ErikEJ** [comment]
@smitpatel Look in this branch, there I have renamed namespace: https://github.com/ErikEJ/EntityFramework.SqlServerCompact/tree/2-1-preview1-prep

**Comment by ErikEJ** [comment]
@ajcvickers  Sounds like a good plan, but agree on the urgency with @roji

**Comment by caleblloyd** [comment]
Thanks for compiling the list, @smitpatel. I agree with the others on the low urgency, we also try to do minor releases quickly that track upstream.

It would be really nice to get to a point where minor and patch releases don't require many changes, because it can be time consuming figuring out what has changed upstream.

**Comment by smitpatel** [comment]
sqlserver

Annotations Framework
	- InternalEntityTypeBuilder
	- ConfigurationSource
	- InternalIndexBuilder
	- InternalModelBuilder
	- InternalPropertyBuilder
	- InternalKeyBuilder
	- InternalRelationalBuilder
	- RelationalForeignKeyBuilderAnnotations
	- RelationalAnnotationsBuilder
	- Property
	- DbFunction
	- Property
	- EntityType
	- CoreAnnotationNames
	- ConventionalAnnotation
	- GetAllBaseTypesInclusive
	- FindPrincipal
	- FindSharedTableRootPrimaryKeyProperty
	- GetDeclaredProperties
	
	
Conventions
	- IModelInitializedConvention
	- RelationalConventionSetbuilder
	- RelationalConventionSetBuilderDependencies
	- RelationalMaxIdentifierLengthConvention
	- ICoventionSetBuilder
	- RelationalDbFunctionConvention
	- IBaseTypeChangedConvention
	- IIndexAddedConvention
	- IIndexUniquenessChangedConvention
	- IIndexAnnotationChangedConvention
	- IPropertyNullabilityChangedConvention
	- IPropertyAnnotationChangedConvention
	- IEntityTypeAnnotationChangedConvention
	- IKeyAddedConvention
	- RelationalValueGeneratorConvention
	- IModelInitializedConvention
	- ValueGeneratorConvention
	- RelationalDbFunctionConvention
	
Query
	- FindProperty(this expression, type)
	- MemberAccessBindingExpressionVisitor.GetPropertyPath
	- QueryCompilationContextDependencies
	- ILinqOperatorProvider
	- LinqOperatorProvider
	- AsyncLinqOperatorProvider
	- LiftExpressionFromSubquery

Logging
	- RelationalStrings
	- CoreStrings
	- AbstractionStrings

Utility
	- DisplayName()
	- ShortDisplayName()
	- NonCapturingLazyInitializer
	- IParameterBindingFactories (referenced in PropertyInfoExtensions)

Migrations/Scaffolding
	- ScaffoldingAnnotationNames
	- IndentedStringBuilder

Update
ISingletonUpdateSqlGenerator

**Comment by smitpatel** [comment]
Sqlite

Annotations Framework
	- InternalEntityTypeBuilder
	- ConfigurationSource
	- InternalIndexBuilder
	- InternalModelBuilder
	- InternalPropertyBuilder
	- InternalKeyBuilder
	- InternalRelationalBuilder
	- RelationalForeignKeyBuilderAnnotations
	- RelationalModelBuilderAnnotations
	- RelationalPropertyBuilderAnnotations
	- RelationalEntityTypeBuilderAnnotations
	- RelationalKeyBuilderAnnotations
	- RelationalIndexBuilderAnnotations
	- GetDeclaredIndexes
	
Conventions
	- RelationalConventionSetbuilder
	- RelationalConventionSetBuilderDependencies
	- ICoventionSetBuilder
	
Query
	- FindProperty(this expression, type)

Logging
	- AbstractionStrings

Utility
	- DisplayName()
	- ShortDisplayName()
	- IParameterBindingFactories (referenced in PropertyInfoExtensions)

Update
	- ISingletonUpdateSqlGenerator

**Comment by smitpatel** [comment]
SqlServerCe

Annotations Framework
	- InternalEntityTypeBuilder
	- ConfigurationSource
	- InternalIndexBuilder
	- InternalModelBuilder
	- InternalPropertyBuilder
	- InternalKeyBuilder
	- InternalRelationalBuilder
	- RelationalForeignKeyBuilderAnnotations
	- RelationalModelBuilderAnnotations
	- RelationalPropertyBuilderAnnotations
	- RelationalEntityTypeBuilderAnnotations
	- RelationalKeyBuilderAnnotations
	- RelationalIndexBuilderAnnotations
	
Conventions
	- RelationalConventionSetbuilder
	- RelationalConventionSetBuilderDependencies
	- CoreConventionSetBuilder
	- CoreConventionSetBuilderDependencies
	- ICoventionSetBuilder
	
Query
	- QueryCompilationContextDependencies
	- ILinqOperatorProvider
	- LinqOperatorProvider
	- AsyncLinqOperatorProvider
	- RemoveConvert

Logging
	- CoreStrings
	- RelationalStrings

Scaffolding
	- TableSelectionSet
	- ScaffoldingAnnotationNames

**Comment by smitpatel** [comment]
Filed #11404, #11405, #11406, #11407, #11408, #11409

**Comment by TechnikEmpire** [comment]
Given that there's super critical fixes in core 2.1 like [this](https://github.com/dotnet/corefx/issues/23401), I'd have to disagree with a lack of urgency. I'm developing against 2.1 precisely because I don't want to exceed the life age of the universe per HTTP request, but my doing so I'm left without a database.

**Comment by ErikEJ** [comment]
@TechnikEmpire  Not sure how your comment relates to this "clean-up" issue?

**Comment by caleblloyd** [comment]
@TechnikEmpire Pomelo maintainer here. This issue is not what is holding up the 2.1-preview. We are an open source project with a small contributor base. We all have jobs and personal lives in addition to time we spend on open source contributions.

EF upstream releases naturally have changes and take work to integrate.

One way to help is reporting issues. Another is contributing code. Pressing issues for higher priority is not constructive though; Pomelo is not an Enterprise maintained project and nobody is getting paid to work on it

**Comment by TechnikEmpire** [comment]
@caleblloyd Hey, guy that doesn't care here. We're not on your repo, you seem to be confused. Next time, before you jump in head first to talk condescendingly to someone, you should check:

 1) did you put your pants on before leaving the house 
 2) do you know what state you're in
 3) do you know what repo you're commenting on

All of your snarky remarks about muh volunteer codez and pressing priority go out the window when you follow steps 1-3 and realize we're on EntityFrameworkCore, not Pomelo. 

I went to submit a PR on your repo. Then I realized that starting from scratch was easier.

By the way high five on muh volunteer coding etc. I'll take one too, hundreds of thousands of changes and additions pushed myself.

**Comment by caleblloyd** [comment]
By no means were my comments meant to be condescending towards any individual, I was more trying to illustrate that this issue is not a blocker for Pomelo.  Pomelo is at a shortage of contributor power right now and code contributions / issues would help that but there's no way to speed it up without more contributors.

Let's stick to the [code of conduct](https://opensource.microsoft.com/codeofconduct/) and leave personal insults out of this.  This is GitHub, not reddit, after all.

**Comment by TechnikEmpire** [comment]
@caleblloyd Don't come around talking to me like you've got your big boy developer pants on and you're going to school me on how to be helpful while patting me on the head like a patronizing goof.

**Comment by divega** [comment]
@TechnikEmpire Reading your first and second comments, I understand your intention was to point out that the EF team (not the Pomelo project) has made what you believe are several mistakes, starting with using .Internal namespaces as a way to indicate that certain APIs should be considered internal rather than using the available internal keyword, and that you believe that this had the effect to encourage actions that we should have tried to discourage.

There were certainly reasons to do things the way we did them but I hope we can still take your feedback and use it to discuss if there are more things we can do around this area in the future (cc @ajcvickers). 

On the other hand, although this issue is closed and the children issue (#11404, #11405, #11406, #11407, #11408, #11409) are in the backlog, we are still planning to mitigate breaks caused by changes in 2.1. This means in some cases we will revert changes, in other cases we will tweak changes to avoid breaking and in some other cases changes will be necessary on providers. We will continue working with provider writers on this. 

Finally, in order to be able to count with your valuable feedback in the future, I need to ask you to follow the code of conduct at https://opensource.microsoft.com/codeofconduct/, in particular be respectful on your interactions with everyone in this repo. All comments you submit count, even if you delete or edit them later.

**Comment by TechnikEmpire** [comment]
@divega 

```
All comments you submit count, even if you delete or edit them later.
```
What a frightening view of the world Microsoft has. What a dark distopian world you espouse, when Microsoft would come talking to it's customers as if it is God Almighty, collecting and keeping every word spoken and judging and condemning based on that. How incredibly frightening indeed. 

Don't you worry Mr. Geheime Staatspolizei sir, I will withdraw from participating in anything Microsoft related and be sure to share your official company world view with others for the greater good. Ein Volk, ein Unternehmen, ein Verhaltensregeln! Microsoft!

**Comment by divega** [comment]
@TechnikEmpire you are welcome to stay and continue participating of this conversation in a civil manner or to leave if you so desire.

By “all comments you submit count”, I was referring to the fact that in my email I found comments from you that I no longer see in this thread. Yes, the code of conduct applies, because everyone that  subscribed to this thread will also receive them.

**Comment by TechnikEmpire** [comment]
@divega aw I'm welcome to stay, thanks buddy. You're giving me permission to participate on the internet. Thank you. So generous. Maybe I should have used English because the reference seemed to have zoomed right by you. Goodbye.

**Comment by roji** [comment]
FWIW as a provider writer I can say that there has been a very good side to exposing internal types as public (under an Internal namespace). It has allowed me to implement functionality even if it involved using internal types, rather than completely blocking on them until the EF team makes them truly public releases a version (or use reflection). That's allowed a very flexible/fast workflow, as long as I was willing to accept some breakage on the way and be reactive. This seems to be a good fit especially for a new product like EF Core where things aren't (or weren't) yet completely nailed down.

It's worth reminding that unenforced encapsulation this is how it works in some other languages (e.g. python has no public/internal/private), it's definitely a legitimate debate.

Regarding the comments (which I also received via email), @TechnikEmpire you seem to have really gone off the deep end here (FYI I'm not part of Microsoft).

**Comment by TechnikEmpire** [comment]
@roji do you feel better tagging me over clearly deleted comments just to have your 2 cents heard? Going "off the deep end" is relative I guess. I view random people who just absolutely cannot resist the narcissistic need to have their opinion heard when no conversation was directed at them as "going off the deep end." Notice how basically everyone else bin'd and ignored the conversation? Not you though. But please by all means, keep smacking those keyboard warrior finger tips.

**Comment by divega** [comment]
@roji I agree re the public/internal surface. This is the main reason we did it. That said, it would have been great if you had to use more specific means to opt-in using an internal type in specific blocks of your code without a warning or error. The way things have been, it is apparently too easy for it to happen accidentally, and for us to miss the fact that an API that we consider is required to implement a provider. Changing namespaces can help a bit, but perhaps a feature more deeply integrated in the language could help more.

**Comment by roji** [comment]
@divega the namespace change is definitely important as we've seen. Maybe simply an analyzer check that produces warnings about usage of EF Core internal code? This way you'd be well-informed and can suppress the warnings if that's what you want. This could be useful for both providers and applications (which can also replace services etc.)

**Comment by divega** [comment]
Yes, an analyzer would be great.

**Comment by ErikEJ** [comment]
Enhance API consistency check and add it to the spec tests?

**Comment by roji** [comment]
@ErikEJ I *think* the problem with doing it in the spec tests is that it's actually OK to use internal code (after all that's why it's public), the main goal is to make sure people are aware that they're doing it. So there needs to be a way to say on specific usages "OK, I'm aware, I want to do it anyway". I'm not sure how such suppression could happen on a per-site basis outside of an analyzer-generated warning.

**Comment by ErikEJ** [comment]
@roji Analyzer suppressions ftw!

**Comment by ajcvickers** [comment]
Consolidated list of remaining work. (Will update individual issues as well, but wanted to have this in one place for me.)


Annotations Framework
- InternalEntityTypeBuilder
- RelationalAnnotationsBuilder
- ~ConfigurationSource~ Public
- InternalIndexBuilder
- InternalModelBuilder
- InternalPropertyBuilder
- InternalKeyBuilder
- InternalRelationalBuilder
- RelationalForeignKeyBuilderAnnotations
- RelationalPropertyBuilderAnnotations
- RelationalIndexBuilderAnnotations
- RelationalKeyBuilderAnnotations
- RelationalEntityTypeBuilderAnnotations
- RelationalAnnotationsBuilder
- Property
- DbFunction
- Property
- EntityType
- ~CoreAnnotationNames~ Public
- ConventionalAnnotation
- ~GetAllBaseTypesInclusive~ Public
- ~FindPrincipal~ Public
- ~FindSharedTableRootPrimaryKeyProperty~ Public
- ~GetDeclaredProperties~ Public
- ~GetDeclaredIndexes~ Public

Conventions
- DatabaseGeneratedAttributeConvention
- IModelInitializedConvention
- RelationalConventionSetbuilder
- RelationalConventionSetBuilderDependencies
- CoreConventionSetBuilder
- CoreConventionSetBuilderDependencies
- IConventionSetBuilder
- RelationalMaxIdentifierLengthConvention
- RelationalMaxIdentifierLengthConvention
- RelationalDbFunctionConvention
- IBaseTypeChangedConvention
- IIndexAddedConvention
- IIndexUniquenessChangedConvention
- IIndexAnnotationChangedConvention
- IPropertyNullabilityChangedConvention
- IPropertyAnnotationChangedConvention
- IEntityTypeAnnotationChangedConvention
- IKeyAddedConvention
- RelationalValueGeneratorConvention
- IModelInitializedConvention
- ValueGeneratorConvention
- RelationalDbFunctionConvention

RevEng
- ~TableSelectionSet~ Removed
- ~ScaffoldingAnnotationNames~ Public

Logging
- ~RelationalStrings~ Public
- ~CoreStrings~ Public
- ~AbstractionStrings~ Public

Utility
- ~DisplayName()~ Public
- ~ShortDisplayName()~ Public
- ~NonCapturingLazyInitializer~ Use LazyInitializer instead
- IParameterBindingFactories (referenced in PropertyInfoExtensions)

Command
- ~IndentedStringBuilder~ No longer needed
- ~RelationalParameterBuilder~ Removed
- ~RelationalCommand~ Public
- ~Logger.CommandExecuting/Executed/Error log mssages~ Public
- ~TransactionLogger.TransactionStarted/Used/Committed/Error/Rolledback~ Public

Query
- ~FindProperty(this expression, type)~ Public
- ~MemberAccessBindingExpressionVisitor.GetPropertyPath~
- QueryCompilationContextDependencies
- ILinqOperatorProvider
- LinqOperatorProvider
- AsyncLinqOperatorProvider
- LiftExpressionFromSubquery
- ~RemoveConvert~ Public

Update
- ~ISingletonUpdateSqlGenerator~ Removed
